### ISSUE TITLE
Local edge agent: fine-tune LFM2.5-1.2B/2.6B-Instruct to answer from a maintained knowledge store, with sources and validation (90% target)

### ISSUE THREAD

**Issue by OriNachum** [agentculture/nvsh#33]
## Goal

A **free, local agent that answers questions on the edge** (Jetson AGX Orin / Thor, DGX Spark, eventually Orin Nano-class boxes) — no hosted API, no network needed at answer time — with **at least 90% of answers correct, each carrying its sources and a validation step**.

The route: fine-tune **`LiquidAI/LFM2.5-1.2B-Instruct`** and **`LiquidAI/LFM2.5-2.6B-Instruct`** not to *memorise* facts, but to be good at **searching and querying an organised local knowledge store**, citing what they found, and checking the answer before giving it. Knowledge lives in the store, where it can be corrected and updated; skill at using the store lives in the weights.

This is future work. It builds on, and is separate from, the tiered local response work (issues #30 and #31: Needle3 Tier 1, LFM2.5 Tier 2, full agent Tier 3).

## Why this shape

- A 1–3B model cannot hold JetPack / L4T / CUDA / container-runtime knowledge reliably, and that knowledge changes with every release. Retrieval keeps answers current without retraining.
- "With sources" is what makes a small model's answer checkable by the operator; "with validation" is what lets the agent decline instead of guessing.
- Measured in the tiers work (2026-09-19, DGX Spark): stock Needle3 picks the right operation on only about half of a 14-request dev corpus, and rewording tool descriptions or reordering tools moves results by noise, not signal. Lesson carried here: **decide the measurement and the data organisation first, then tune** — small models need task-specific training and a disciplined corpus, not prompt tweaks.

## Deliverables

1. **Knowledge store** — a local, inspectable store (files + an index; no server required) holding, at minimum: nvsh's own docs, platform facts per device (detected values and their sources, as in `docs/platforms.md`), NVIDIA/Jetson setup and troubleshooting knowledge, and the operator's own machine notes. Every entry has: a stable id, source (URL or file + revision), date, the platforms/versions it applies to, and a confidence/verification status.
2. **Query interface the model is trained on** — a small, typed set of read-only tools (for example `search(query, filters)`, `get(id)`, `list_sources(topic)`, plus the existing nvsh inspection operations for live machine state). Typed and table-driven like `nvsh/ops`, so adding a tool or a source does not need a redesign.
3. **Maintenance tools** — commands to add, update, deprecate and verify entries; detect stale entries (source changed, version superseded); rebuild the index; and report coverage gaps found from unanswered or wrongly answered questions. Redaction applies to anything derived from the operator's machine.
4. **Training data and recipes** — a question/answer corpus where each item records the question, the tool calls a good answer needs, the supporting entry ids, the answer, and a should-decline flag; split into dev and held-out, with held-out authored separately from the store's text. Recipes for unsloth, unsloth-cli and a remote Hugging Face run; published under `jetson-ai-lab` with the agreed naming (`jetson-ai-lab/lfm2.5-1.2b-nvsh-<task>`, `…-2.6b-…`, `-GGUF` builds, dataset repo alongside), tags `ops<N>-r<M>`-style versioning tied to the tool/schema generation.
5. **Evaluation harness** — extends `nvsh tiers bench`: per-item rows, deterministic scoring, provenance (model hashes, device, engine), and the targets below.

## Success criteria (measured on held-out questions, on-device)

- **≥ 90% answers correct**, where "correct" requires: the answer is right, **every claim is backed by a cited store entry that actually supports it**, and the validation step passed.
- **Unsupported-claim rate ≤ 2%** (an answer citing a source that does not say that counts as wrong).
- **Should-decline recall ≥ 90%**: when the store does not contain the answer, the agent says so (and may escalate to the configured full agent) instead of inventing one.
- Works **offline**; memory and latency budgets stated per device for both model sizes, so the 1.2B / 2.6B trade-off is a measured choice.
- nvsh invariants hold: the agent proposes, never runs a mutating command without operator approval; nothing leaves the box unless the operator exports it.

## Open questions to settle first

- Store format and index: plain files + a lexical index (stdlib-friendly) versus embeddings (needs a runtime and memory) — or lexical first, embeddings as an optional flavor.
- What "validation" is, concretely: a second pass that checks each claim against the cited entry, a live inspection operation that confirms a machine-specific claim, or both.
- The **LFM Open License** terms for publishing fine-tuned derivatives (must be read before any tuned LFM is published).
- Which inference engine per device (llama.cpp / vLLM / SGLang) once memory is measured on an Orin-class box.
- Where the question corpus comes from without leaking operator data: authored, derived from public docs/forums with attribution, and reviewed redacted exports.

## Suggested first step

Run this through the devague method (`/think` → `/challenge` → `/spec-to-plan`) as its own frame, starting from the success criteria above; build the evaluation harness and a small store **before** any training run, and record a stock-model baseline for both sizes.

- nvsh (Claude)

**Comment by OriNachum** [comment]
## Design input relayed by the operator (suggested by Gemini), with how it fits

Three suggestions, all consistent with the issue's direction ("skill in the weights, knowledge in the store"). Recording them here so the `/think` pass starts from them.

### 1. Pre-index into hierarchical schemas, not free-form dumps

Do not make the model invent fuzzy search strings. Give it a **strict schema with enumerated, structured arguments**, and index the store along those same axes:

```json
{
  "name": "search_jetson_docs",
  "parameters": {
    "module": ["camera_csi", "pcie", "tensorrt", "docker_containers", "gpio", "power_modes"],
    "jetpack_version": ["JP5", "JP6", "any"],
    "query": "exact keywords or symbol"
  }
}
```

Fit: this is deliverable 2 (typed query interface) made concrete, and it settles part of the first open question — a **faceted lexical index** (module x platform x version, then keywords) is enough to start, needs no embedding runtime, and stays within an Orin-class memory budget. The enumerations must be **generated from the store's own taxonomy**, the way nvsh generates Needle's tool schemas from `nvsh/ops/table.py`, so adding a module or a JetPack release is a data change, not a code or retraining-format change. A small model choosing from an enum is also directly measurable (facet accuracy) separately from answer quality.

### 2. Train on "tool first, synthesize second"

Fine-tune (SFT, then DPO) on trajectories where the model **always defers factual lookups to a tool** and only then writes the answer:

```text
Input:        How do I allocate 15W nvpmodel on AGX Orin?
Tool call:    {"name": "get_power_mode", "device": "agx_orin", "mode": "15W"}
Observation:  {"supported": false, "lowest_mode": "power_mode_0_20W"}
Final answer: 15W is not a stock preset on AGX Orin; the closest profile is ... (cites the entry)
```

Fit: this is the corpus shape in deliverable 4 (question -> required tool calls -> supporting entry ids -> answer). Two additions it implies: (a) DPO pairs where the **rejected** sample answers from memory without a tool call, or cites an entry that does not support the claim — that trains the behaviour the 2% unsupported-claim target measures; (b) observations that say "not supported / not found" must be common in the data, so declining is a learned, normal outcome (the should-decline target).

### 3. Hybrid alignment (knowledge-guided routing)

Teach the model the **domain vocabulary** so it knows *what to look up* — for example that "GStreamer pipeline crash with IMX219" points at `nvarguscamerasrc` and device-tree overlays — while code snippets and parameter values stay in the tool's return.

Fit: this draws the line for what may live in the weights: **routing knowledge** (which module/facet a symptom belongs to, which terms are synonyms) yes; **values** (flags, versions, paths, wattages) no — those go stale per release and must come from a cited entry. It also gives the maintenance tools a job: when a question is mis-routed, the fix is a taxonomy/synonym entry plus a training example, and the harness should report mis-routes separately from wrong answers.

### Consequences for the evaluation harness

Score three things separately, per item: **routing** (right tool and facets), **grounding** (every claim supported by the cited entry), **answer** (correct and validated). The 90% target applies to the conjunction; the split tells us whether to fix the taxonomy, the store, or the model.

- nvsh (Claude)

**Comment by OriNachum** [comment]
## Candidate component: Laya (a non-generative decision model)

Looked at on 2026-09-19 from its product page only (`laya.convaiinnovations.com`); nothing was run, so every figure below is the publisher's claim.

**What it is.** An encoder model (ModernBERT-large, 421M parameters; a multilingual mmBERT-base variant at 322M) that does not generate text. It answers typed questions over text or JSON with calibrated probabilities, through three primitives: `choice` (pick one of a fixed set), `score` (ordinal) and a boolean. Apache-2.0, weights on Hugging Face (`convaiinnovations/laya`), a PyPI SDK. Claimed 32.8 ms per decision on a single GPU, 7.2 ms per question batched; context 512 to 1024 tokens (8k for the multilingual one). Tested on T4; no Jetson documentation.

**Where it could fit this issue's agent.**

- **Routing before retrieval** ("knowledge-guided routing" in the first comment): a `choice` over the knowledge store's top-level modules is exactly the structured, enumerated decision suggested there, and a calibrated one.
- **Answer validation** (the "90% success with sources and validations" goal): a boolean "does this source passage support this answer?" with a calibrated probability is a cheap validator to put behind the LFM2.5 generator, and a usable abstain signal.
- **Deciding when not to answer.** In the tiers work (PRs #32, #34 and the Tier 2 PR) the weakest measured behaviour of every small model was declining: stock Needle3 declines 74% of should-decline requests, a LoRA-tuned Needle3 16 of 27, stock LFM2.5-350M 1 of 27. A dedicated classifier for "is this in scope" is the obvious thing to try.

**Where it does not fit.** It cannot be the answering model or a tool-calling agent: no generation, no argument extraction.

**Cost to weigh.** It needs PyTorch and transformers and about 0.8 GB at fp16, against a 35 MB CPU-only engine for Needle3; on a Jetson that means a container or a GPU PyTorch build. Footprint is a first-class constraint here, so measure resident memory and CPU-only latency on an Orin before adopting it.

**Suggested first step.** A spike on the DGX Spark that scores Laya as (a) the should-decline gate and (b) the yes/no check on a proposed operation, against nvsh's 318-entry grid corpus (`nvsh/tiers/corpus/dev.json`, 106 should-decline entries), beside the numbers already recorded in `docs/tiers-improving-accuracy.md`.

- nvsh (Claude)
